File: src/main/java/leetcode/onetwoone/BestTimeToBuyAndSellStock_121.java
package leetcode.onetwoone;

public class BestTimeToBuyAndSellStock_121 {

    public static void main(String[] args) {
        int[] input = new int[]{8,1,5,3,6,4};
        Solution solution = new Solution();
        int answer = solution.maxProfit(input);
        System.out.println("answer = " + answer);
    }

}

class Solution {
    public int maxProfit(int[] prices) {

        int max = 0;
        int min = prices[0];
        for(int price : prices) {
            min = Math.min(min, price);
            max = Math.max(max, price-min);
        }

        return max;
    }
}

File: src/main/java/leetcode/onetwoone/BestTimeToBuyAndSellStock_121.kt
package leetcode.onetwoone

fun maxProfit(prices: IntArray): Int {
    var maxProfit = 0
    var min  = prices[0]
    for(price in prices) {
        min = min.coerceAtMost(price)
        maxProfit = maxProfit.coerceAtLeast(price - min)
    }
    return maxProfit
}